### PR TITLE
AGS: Support playing Maniac Mansion Deluxe without uncompress/install

### PR DESCRIPTION
**Pull request by phcoder**

### CODE REVIEW

File: common/clickteam.cpp [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+	for (p = strings; p < (char*)tag + lmax && *p; p++);
+	_fileName = Common::String(strings, p - strings);
+	_fileDescriptorOffset = off;
+}
+
+ClickteamInstaller::ClickteamTag* ClickteamInstaller::getTag(ClickteamTagId tagId) const {
+	return _tags.getValOrDefault((uint16) tagId).get();
+}
+
+static uint32 computeCRC(byte* buf, uint32 sz, unsigned int previous)
```
Comment (sev-) [inline common/clickteam.cpp:71]: Consider using `uint` here for brevity. Also, `byte *buf`

File: common/clickteam.cpp [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+{
+	uint32 cur = previous;
+	byte *ptr = buf;
+	uint32 i;
+
+	for (i = 0; i < (sz & ~3); i+=4, ptr += 4)
```
Comment (sev-) [inline common/clickteam.cpp:77]: `i += 4`. etc/ Please, review the whole thing

File: common/clickteam.cpp [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+	_fileName = Common::String(strings, p - strings);
+	_fileDescriptorOffset = off;
+}
+
+ClickteamInstaller::ClickteamTag* ClickteamInstaller::getTag(ClickteamTagId tagId) const {
+	return _tags.getValOrDefault((uint16) tagId).get();
+}
+
+static uint32 computeCRC(byte* buf, uint32 sz, unsigned int previous)
+{
```
Comment (sev-) [inline common/clickteam.cpp:72]: the code formatting is off in many places here.  Please, check how the curly braces stick in other instances too.

File: common/clickteam.cpp [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+	}
+
+	if (block3_offset <= 0 || block3_len <= 0)
+		return nullptr;
+
+	return new ClickteamInstaller(files, caseMap, tags, crc_xor, block3_offset, block3_len, stream, dispose);
+}
+
+static Common::String translateName(const Path &path) {
+	Common::String lower = path.toString('\\');
```
Comment (sev-) [inline common/clickteam.cpp:226]: You could use '/' as a separator and then call `Common::normalizePath()` which is doing exactly the same thing.

File: common/clickteam.h [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+	ClickteamInstaller(Common::HashMap<String, ClickteamFileDescriptor> files,
+			   Common::HashMap<Common::String, Common::String> caseMap,
+			   Common::HashMap<uint16, Common::SharedPtr<ClickteamTag>> tags,
+			   uint32 crcXor, uint32 block3Offset, uint32 block3Size, Common::SeekableReadStream *stream, DisposeAfterUse::Flag dispose)
+		: _files(files), _caseMap(caseMap), _tags(tags), _crcXor(crcXor), _block3Offset(block3Offset), _block3Size(block3Size), _stream(stream, dispose) {
+	}
+
+	Common::HashMap<Common::String, ClickteamFileDescriptor> _files;
+	Common::HashMap<Common::String, Common::String> _caseMap;
```
Comment (sev-) [inline common/clickteam.h:93]: Use `Common::StringMap` here, that is also ignoring case on the comparison, so you do not need to lowercase your paths.

File: engines/ags/ags.cpp [hunk 6]
```diff
@@ -70,6 +71,17 @@
 
 namespace AGS {
 
+struct AGSInstallerTableEntry {
+	const char *installerName;
+	const char *mainFileName;
+};
+
+static const AGSInstallerTableEntry agsInstallerTable[] = {
+	{"manicmdsetup.exe", "Maniac.exe"},
```
Comment (sev-) [inline engines/ags/ags.cpp:80]: I would personally stick it into `extra` in the detection and a game flag (GF_). Something like

```
GAME_ENTRY_INSTALLER(maniacmansiondeluxe", "manicmdsetup.exe",  "Maciac.exe", "329ff478ffc1a3bd0dc941d5212aa602", 5826271),

// And that specifies GF_INSTALLER
```

But that is a matter of taste. My reasoning is to keep it in a single place, so things do not get desync accidentally.

File: engines/ags/ags.cpp [hunk 7]
```diff
@@ -152,6 +164,15 @@ Common::Error AGSEngine::run() {
 	setDebugger(new AGSConsole(this));
 
 	const char *filename = _gameDescription->desc.filesDescriptions[0].fileName;
+	for (unsigned i = 0; i < ARRAYSIZE(agsInstallerTable); i++) {
```
Comment (sev-) [inline engines/ags/ags.cpp:167]: `uint` is so clean and short ;)

**Review by sev- — CHANGES_REQUESTED** [review]
Looks good besides small formatting suggestions and a couple of open questions.

File: common/clickteam.cpp [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+
+#define STUB_SIZE 0x16000
+#define FLAG_COMPRESSED 1
+
+namespace Common {
+
+ClickteamInstaller::ClickteamFileDescriptor::ClickteamFileDescriptor(const ClickteamTag& contentsTag, uint32 off)
+  : _fileDataOffset(0), _fileDescriptorOffset(0), _compressedSize(0), _uncompressedSize(0) {
+	uint32 stringsOffset = 36;
+	unsigned char *tag = contentsTag._contents + off;
```
[the remark this answers is not in the document]
Comment (phcoder) [reply common/clickteam.cpp:38]: Done

File: common/clickteam.cpp (hunk 1 above)
Comment (phcoder) [reply common/clickteam.cpp:71]: Done

File: common/clickteam.cpp (hunk 2 above)
Comment (phcoder) [reply common/clickteam.cpp:77]: Done

File: common/clickteam.cpp (hunk 3 above)
Comment (phcoder) [reply common/clickteam.cpp:72]: Done

File: common/clickteam.cpp (hunk 4 above)
Comment (phcoder) [reply common/clickteam.cpp:226]: Done while using sep = '\\' argument to normalizePath. Archive uses \ internally, so it's simpler without translation

File: common/clickteam.h (hunk 5 above)
Comment (phcoder) [reply common/clickteam.h:93]: Done

File: engines/ags/ags.cpp (hunk 7 above)
Comment (phcoder) [reply engines/ags/ags.cpp:167]: Done

File: engines/ags/ags.cpp (hunk 6 above)
Comment (phcoder) [reply engines/ags/ags.cpp]: Done

**Comment by phcoder** [comment]
Changes done. PTAL

**Comment by sev-** [comment]
Excellent!

**Pull request merged**